### PR TITLE
seperate traffic manager endpoint

### PR DESCRIPTION
**Pull request by yungezz**
seperate traffic manager endpoint into new module

### CODE REVIEW

File: library/azure_rm_trafficmanagerendpoint.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+    def exec_module(self, **kwargs):
+
+        for key in list(self.module_arg_spec.keys()):
+            setattr(self, key, kwargs[key])
+
+        to_be_updated = False
+
+        if not self.location:
+            self.location = 'global'
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: traffic manager endpoint can have location like 'WestUS'. I tested use 'global' as endpoint_location and got error msg that global is invalid.

File: library/azure_rm_trafficmanagerendpoint.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+                type='str',
+                choices=['azureEndpoints', 'externalEndpoints', 'nestedEndpoints'],
+                required=True
+            ),
+            target_resource_id=dict(type='str'),
+            enable=dict(type='bool', default=True),
+            weight=dict(type='int'),
+            priority=dict(type='int'),
+            location=dict(type='str'),
+            minitor_status=dict(type='str'),
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: Endpoint monitor status cannot be changed manually. It's a combination of the results of endpoint monitoring and the configured endpoint status. link for ref: https://docs.microsoft.com/en-us/azure/traffic-manager/traffic-manager-monitoring#endpoint-monitor-status
It also should not be check in check_update()

File: library/azure_rm_trafficmanagerendpoint.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        self.state = None
+
+        self.profile_name = None
+        self.type = None
+        self.target_resource_id = None
+        self.enable = None
+        self.weight = None
+        self.priority = None
+        self.location = None
+        self.minitor_status = None
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: monitor_status

File: library/azure_rm_trafficmanagerendpoint.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+        if self.enable is not None and self.enable == False:
+            self.endpoint_status = 'Disabled'
+
+        response = self.get_traffic_manager_endpoint()
+
+        if response:
+            self.log('Results : {0}'.format(response))
+            if self.state == 'present':
+                # check update
+                to_be_update = selff.check_update(response)
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: typo self

File: library/azure_rm_trafficmanagerendpoint.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+        '''
+        self.log("Creating / Updating the Traffic Manager endpoint {0}".format(self.name))
+
+        parameters = Endpoint(target_resource_id=self.target_resource_id,
+                              target=self.target,
+                              endpoint_status=self.endpoint_status,
+                              weight=self.weight,
+                              priority=self.priority,
+                              endpoint_location=self.location,
+                              endpoint_monitor_status=self.monitor_status,
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: monitor_status

File: library/azure_rm_trafficmanagerendpoint.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+        if (response['weight'] != self.weight:
+            self.log("weight Diff - Origin {0} / Update {1}".format(response['weight'], self.weight))
+            return True
+
+        if (response['priority'] != self.priority:
+            self.log("priority Diff - Origin {0} / Update {1}".format(response['priority'], self.priority))
+            return True
+
+        if (response['monitor_status'] != self.monitor_status:
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint.py]: monitor status

File: library/azure_rm_trafficmanagerendpoint_facts.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+short_description: Get Azure Traffic Manager endpoint facts
+
+description:
+    - Get facts for a specific Traffic Manager endpoints or all endpoints  in a Traffic Manager profile
+
+options:
+    name:
+        description:
+            - Limit results to a specific Traffic Manager profile.
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint_facts.py]: specific Traffic Manager endpoint

File: library/azure_rm_trafficmanagerendpoint_facts.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+        type=endpoint.type,
+        target_resource_id=endpoint.target_resource_id,
+        target=endpoint.target,
+        endpoint_status=endpoint.endpoint_status,
+        weight=endpoint.weight,
+        priority=endpoint.priority,
+        endpoint_location=endpoint.endpoint_location,
+        min_child_endpoints=endpoint.min_child_endpoints,
+        geo_mapping=endpoint.geo_mapping,
+    )
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint_facts.py]: monitor_status can be part of facts

File: library/azure_rm_trafficmanagerendpoint_facts.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+        if self.name and not self.resource_group:
+            self.fail("Parameter error: resource group required when filtering by name.")
+
+        if self.name:
+            self.results['tms'] = self.get_item()
+        elif self.resource_group:
+            self.results['tms'] = self.list_resource_group()
+        else:
+            self.results['tms'] = self.list_all()
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint_facts.py]: self.results['endpoints']

File: library/azure_rm_trafficmanagerendpoint_facts.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+        return self.results
+
+    def get_item(self):
+        """Get a single Azure Traffic Manager"""
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint_facts.py]: Get a single Azure Traffic Manager Endpoint

File: library/azure_rm_trafficmanagerprofile.py [hunk 11]
```diff
@@ -418,13 +316,14 @@ def __init__(self):
...
         self.dns_config = None
         self.monitor_config = None
-        self.endpoints = None
 
         self.results = dict(
             changed=False
         )
 
-        super(AzureRMTrafficManager, self).__init__(derived_arg_spec=self.module_arg_spec, supports_check_mode=True)
+        super(AzureRMTrafficManagerProfiles, self).__init__(derived_arg_spec=self.module_arg_spec,
```
Comment (caohai) [inline library/azure_rm_trafficmanagerprofile.py]: AzureRMTrafficManagerProfiles->AzureRMTrafficManagerProfile

File: library/azure_rm_trafficmanagerprofile.py [hunk 12]
```diff
@@ -525,8 +424,7 @@ def ceate_update_traffic_manager_profile(self):
             profile_status=self.profile_status,
             traffic_routing_method=self.traffic_routing_method,
             dns_config=create_dns_config_instance(self.dns_config) if self.dns_config else None,
-            monitor_config=create_monitor_config_instance(self.monitor_config) if self.monitor_config else None,
-            endpoints=create_endpoints(self.endpoints)
```
Comment (caohai) [inline library/azure_rm_trafficmanagerprofile.py]: Not sure about this one. If no value is passed to endpoints when creating a Profile object, will updating the profile delete all endpoints?

File: library/azure_rm_trafficmanagerprofile.py [hunk 13]
```diff
@@ -558,22 +456,12 @@ def check_update(self, response):
...
-                        if str(v).lower() != str(e2[k]).lower():
-                            self.log("Endpoints Diff - Origin {0} / Update {1}".format(response['endpoints'], self.endpoints))
-                            return True
         return False
 
 
 def main():
     """Main execution"""
-    AzureRMTrafficManager()
+    AzureRMTrafficManagerProfiles()
```
Comment (caohai) [inline library/azure_rm_trafficmanagerprofile.py]: AzureRMTrafficManagerProfile

File: library/azure_rm_trafficmanagerprofile.py [hunk 14]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/python
```
Comment (caohai) [inline library/azure_rm_trafficmanagerprofile.py]: I also made some changes in ansible/ansible, mainly docs and case problem.

File: tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml [hunk 15]
```diff
@@ -160,23 +117,136 @@
...
+      - facts.endpoints[0].name == "{{ endpointname1 }}"
+
+- name: Create second Traffic Manager endpoint
+  azure_rm_trafficmanagerendpoint:
+    resource_group: "{{ resource_group }}"
+    profile_name: "{{ tmname }}"
+    name: "{{ endpointname2 }}"
+    type: externalEndpoints
+    location: westus
+    priority: `
```
Comment (caohai) [inline tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml]: typo?

File: tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml [hunk 16]
```diff
@@ -160,23 +117,136 @@
...
+    profile_name: "{{ tmname }}"
+  register: facts
+
+- name: Check 1 endpoint left in profile
+  assert:
+    that:
+      - facts.endpoints.length == 1
+
 - name: Delete the Traffic Manager profile(check mode)
   azure_rm_trafficmanager:
```
Comment (caohai) [inline tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml]: module name need to be updated

File: library/azure_rm_trafficmanagerendpoint_facts.py [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+
+extends_documentation_fragment:
+    - azure
+
+author:
+    - "Hai Cao <t-haicao@microsoft.com>"
+    - "Yunge Zhu <yungez@microsoft.com>"
+'''
+
+EXAMPLES = '''
```
Comment (caohai) [inline library/azure_rm_trafficmanagerendpoint_facts.py]: EXAMPLES and RETURN need to be updated

File: library/azure_rm_trafficmanagerendpoint.py (hunk 5 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerendpoint.py]: fixed

File: library/azure_rm_trafficmanagerendpoint_facts.py (hunk 8 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerendpoint_facts.py]: fixed

File: library/azure_rm_trafficmanagerendpoint_facts.py (hunk 9 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerendpoint_facts.py]: fixed

File: library/azure_rm_trafficmanagerprofile.py (hunk 11 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerprofile.py]: fixed

File: library/azure_rm_trafficmanagerprofile.py (hunk 12 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerprofile.py]: no, should pass what  get to the create_or_update

File: tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml (hunk 15 above)
Comment (yungezz) [reply tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml]: fixed

File: tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml (hunk 16 above)
Comment (yungezz) [reply tests/integration/targets/azure_rm_trafficmanagerendpoint/tasks/main.yml]: fixed

File: library/azure_rm_trafficmanagerendpoint_facts.py (hunk 17 above)
Comment (yungezz) [reply library/azure_rm_trafficmanagerendpoint_facts.py]: fixed

File: library/azure_rm_trafficmanagerendpoint.py [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+    profile_name:
+        description: Name of Traffic Manager profile where this endpoints attaches to.
+        type: str
+        required: true
+    type:
+        description:
+            - The type of the endpoint.
+        required: true
+        choices:
+            - azureEndpoints
```
Comment (zikalino) [inline library/azure_rm_trafficmanagerendpoint.py]: choices sound a bit strange in plural, and also no need to keep "Endpoint" in the name. Maybe should be:
- azure
- external
- nested

File: library/azure_rm_trafficmanagerendpoint.py [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        self.log("Deleting the Traffic Manager endpoint {0}".format(self.name))
+        try:
+            operation_result = self.traffic_manager_management_client.endpoints.delete(self.resource_group, self.profile_name, self.type, self.name)
+            return True
+        except CloudError as exc:
+            request_id = exc.request_id if exc.request_id else ''
+            self.fail("Error deleting the Traffic Manager endpoint {0}, request id {1} - {2}".format(self.name, request_id, str(exc)))
+            return False
+
+    def ceate_update_traffic_manager_endpoint(self):
```
Comment (zikalino) [inline library/azure_rm_trafficmanagerendpoint.py]: ceate -> create

Comment (yungezz) [reply library/azure_rm_trafficmanagerendpoint.py]: fixed

**Pull request closed without merging**